### PR TITLE
Add cases STRESSTEST-VERIFY-RESTART-with multiple scenarios.

### PR DESCRIPTION
**Pull request by LiliDeng**
Add test cases STRESSTEST-VERIFY-RESTART-MAX-NVME-DISK, STRESSTEST-VERIFY-RESTART-MAX-GPU-GRID and STRESSTEST-VERIFY-RESTART-MAX-GPU-CUDA.

Test results -
```
[LISAv2 Test Results Summary]
Test Run On           : 04/09/2020 09:53:47
ARM Image Under Test  : SUSE : sles-12-sp5 : gen1 : latest
Initial Kernel Version: 4.12.14-16.7-azure
Final Kernel Version  : 4.12.14-16.7-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:8

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART-MAX-GPU-CUDA                                            PASS                 4.77 

[LISAv2 Test Results Summary]
Test Run On           : 04/09/2020 08:50:24
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.0.0-1035-azure
Final Kernel Version  : 5.0.0-1035-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:16

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART-MAX-SRIOV-NICS                                          PASS                11.73 

[LISAv2 Test Results Summary]
Test Run On           : 04/09/2020 08:28:33
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.0.0-1035-azure
Final Kernel Version  : 5.0.0-1035-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:16

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART-MAX-NVME-DISK                                           PASS                 8.81 

[LISAv2 Test Results Summary]
Test Run On           : 04/10/2020 01:58:49
ARM Image Under Test  : canonical : ubuntuserver : 16.04-lts : Latest
Initial Kernel Version: 4.15.0-1075-azure
Final Kernel Version  : 4.15.0-1075-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:13

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 GPU                  NVIDIA-CUDA-DRIVER-VALIDATION-MAX-GPU                                             PASS                 10.6 
	Using nVidia driver : CUDA 
	lsvmbus: Expected "PCI Express pass-through" count: 4, count inside the VM: 4 : PASS 
	lspci: Expected "3D controller: NVIDIA Corporation" count: 4, found inside the VM: 4 : PASS 
	lshw: Expected Display adapters: 4, total adapters found in VM: 4 : PASS 
	nvidia-smi: Expected GPU count: 4, found inside the VM: 4 : PASS 

[LISAv2 Test Results Summary]
Test Run On           : 04/09/2020 09:58:56
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.0.0-1035-azure
Final Kernel Version  : 5.0.0-1035-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:2:31

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART                                                         PASS               149.65 

[LISAv2 Test Results Summary]
Test Run On           : 04/10/2020 13:29:15
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.0.0-1035-azure
Final Kernel Version  : 5.0.0-1035-azure
Total Test Cases      : 2 (2 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:24

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART                                                         PASS                 1.78 
    2 STRESS               STRESSTEST-VERIFY-RESTART-MAX-GPU-GRID                                            PASS                17.42 
```

### CODE REVIEW

File: Libraries/CommonFunctions.psm1 [hunk 1]
```diff
@@ -2341,3 +2341,40 @@ Function Get-DeviceName
         -command "bash get_data_disk_dev_name.sh" -runAsSudo
     return $ret
 }
+
+Function Get-ExpectedCount {
```
Comment (harz566) [inline Libraries/CommonFunctions.psm1]: Change function Name to a meaningful  better name, adding something like vCPU....

File: Libraries/CommonFunctions.psm1 [hunk 2]
```diff
@@ -2341,3 +2341,40 @@ Function Get-DeviceName
...
+    )
+    $vmCPUCount = Run-LinuxCmd -username $user -password $password -ip $ip -port $port -command "nproc" -ignoreLinuxExitCode
+    if ($vmCPUCount) {
+        Write-LogDbg "Successfully fetched nproc result: $vmCPUCount"
+    } else {
+        Write-LogErr "Could not fetch the nproc command result."
+    }
+    [int]$expectedCount = 0
+    switch ($type) {
+        {($_ -imatch "NVME")} {
```
Comment (harz566) [inline Libraries/CommonFunctions.psm1]: can be just like below, as it's more straight forward, and fitting your functionality design.
GPU {... ;break;}
NVME {...; break;}

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 3]
```diff
@@ -17,54 +17,107 @@ function Main {
...
-            }
-        } catch {
-            $ErrorMessage =  $_.Exception.Message
-            $ErrorLine = $_.InvocationInfo.ScriptLineNumber
-            Write-LogInfo "EXCEPTION : $ErrorMessage at line: $ErrorLine"
+    $expectedcount = 0
+    switch ($testType) {
+        {($_ -imatch "NVME")} {
+            $keyWord = "Non-Volatile memory controller";
+            $expectedcount = Get-ExpectedCount -ip $AllVmData.PublicIP -port $AllVmData.SSHPort -username $user -password $password -size $AllVmData.InstanceSize -type "NVME"
```
Comment (harz566) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: We can use just one line of '$output= Get-ExpectedCount -ip .....',  then split $output into $expectedCount and $keyWord.  Yes, modify the common function Get-ExpectedCount
Remember we should follow a proper naming rule, the local variable should be 'camelCase', so $expectedCount not $expectedcount, just like $keyWord

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 4]
```diff
@@ -17,54 +17,107 @@ function Main {
...
+    } else {
+        for ($rebootNr = 1; $rebootNr -le $rebootNumber; $rebootNr++) {
+            try {
+                Write-LogInfo ("Trying to restart {0}: {1} / {2} ..." `
+                    -f @($AllVMData.RoleName, $rebootNr, $rebootNumber))
+                $isRestarted = $TestProvider.RestartAllDeployments($AllVmData)
+                if ($isRestarted) {
+                    Write-LogInfo "Virtual machine restart successful."
+                    $noIssues = Check-KernelLogs -allVMData $AllVmData
+                    if (!$noIssues) {
```
Comment (harz566) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: rename the variable like "$issues" or "hasIssues", as it's not intuitive and  reader friendly, especially using a condition to do logic if ... else...,

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 5]
```diff
@@ -17,54 +17,107 @@ function Main {
...
+                        break
+                    } else {
+                        if ($testType) {
+                            $pciDevices = Run-LinuxCmd -ip $AllVmData.PublicIP -port $AllVmData.SSHPort -username $user -password $password -command "lspci | grep -i '$keyWord' | wc -l" -runAsSudo -ignoreLinuxExitCode
+                            Write-LogInfo "Actual count: $pciDevices, expected count: $expectedcount"
+                            if ($pciDevices -ne $expectedcount) {
+                                $CurrentTestResult.TestSummary += New-ResultSummary -testResult "FAIL" `
+                                    -metaData "The $rebootNr Reboot: lspci return values are not expected" -checkValues "PASS,FAIL,ABORTED" -testName $currentTestData.testName
+                                $testResult = "FAIL"
+                                break
```
Comment (harz566) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1:68]: do we really want 'break' from all result collection, is it better to collect all of those Failure reason from all restart iterations?

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 6]
```diff
@@ -17,54 +17,107 @@ function Main {
...
+                    $testResult = "FAIL"
+                    $CurrentTestResult.TestSummary += New-ResultSummary -testResult "FAIL" `
+                        -metaData "The $rebootNr Reboot fail" -checkValues "PASS,FAIL,ABORTED" -testName $currentTestData.testName
+                    break
+                }
+            } catch {
+                $ErrorMessage =  $_.Exception.Message
+                $ErrorLine = $_.InvocationInfo.ScriptLineNumber
+                Write-LogInfo "EXCEPTION : $ErrorMessage at line: $ErrorLine"
+                break
```
Comment (harz566) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1:83]: as above

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 (hunk 5 above)
Comment (LiliDeng) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1:67]: When issue happen, we want to keep the site, and see what happen at that time?

Comment (LiliDeng) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1:68]: > do we really want 'break' from all result collection, is it better to collect all of those Failure reason from all restart iterations?

When issue happen, we want to keep the site, and see what happen at that time?

**Comment by LiliDeng** [comment]
@harz566 Thanks, please help review again, then I will do further fix or run tests against latest changed code.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 (hunk 5 above)
Comment (harz566) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1:68]: it's ok if our design just wants to catch the first failure during rebooting and deployments, break and skip others.  If our design is try to see if another reboot can overcome some random errors, then use continue.  I'm not stick to testing logics. both seems ok.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 7]
```diff
@@ -17,54 +17,85 @@ function Main {
...
-            if ($isRestarted) {
-                Write-LogInfo "Virtual machine restart successful."
-                $testResult = "PASS"
-            } else {
-                Write-LogErr "Virtual machine restart failed."
-                $testResult = "FAIL"
+    $expectedCount,$keyWord = Get-ExpectedDevicesCount -vmData $AllVmData -username $user -password $password -type $testType
+    Run-LinuxCmd -ip $AllVmData.PublicIP -port $AllVmData.SSHPort -username $user -password $password -command "which lspci || (. ./utils.sh && install_package pciutils)" -runAsSudo | Out-Null
+    $hasIssues = Check-KernelLogs -allVMData $AllVmData
+    if (!$hasIssues) {
```
Comment (harz566) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: if ($hasIssues) {...}  is it better to have no '!' here?

File: Libraries/CommonFunctions.psm1 [hunk 8]
```diff
@@ -2341,3 +2341,51 @@ Function Get-DeviceName
...
+Function Get-ExpectedDevicesCount {
+    param (
+        [Object] $vmData,
+        [String] $username,
+        [String] $password,
+        [String] $type
+    )
+    $vmCPUCount = Run-LinuxCmd -username $username -password $password -ip $vmData.PublicIP -port $vmData.SSHPort -command "nproc" -ignoreLinuxExitCode
+    $size = $vmData.InstanceSize
+    $keyWord = ""
```
Comment (harz566) [inline Libraries/CommonFunctions.psm1]: probably, no need to define initial value with empty string here.

**Review by harz566 — APPROVED** [review]

**Review by juhlee-microsoft — APPROVED** [review]
This is not the gateway here but I recommend you run those new stress tests with higher counts like 100 or 500 for sustainability.

**Comment by juhlee-microsoft** [comment]
@LiliDeng why is it still 'do-not-merge'? Are you running any testing?

**Comment by LiliDeng** [comment]
> @LiliDeng why is it still 'do-not-merge'? Are you running any testing?

Thanks Ju. I am running tests with more higher counts, and see how long it will take, and there are also some places need improve.

File: Libraries/CommonFunctions.psm1 [hunk 9]
```diff
@@ -2341,3 +2341,66 @@ Function Get-DeviceName
...
+    if ($vmCPUCount) {
+        Write-LogDbg "Successfully fetched nproc result: $vmCPUCount"
+    } else {
+        Write-LogErr "Could not fetch the nproc command result."
+    }
+    [int]$expectedCount = 0
+    switch ($type) {
+        "NVME" {
+            [int]$expectedCount = $($vmCPUCount/8)
+            $keyWord = "Non-Volatile memory controller"
```
Comment (juhlee-microsoft) [inline Libraries/CommonFunctions.psm1:2379]: Isn't it nice to print out keyWord and expectedCount?

File: TestProviders/AzureProvider.psm1 [hunk 10]
```diff
@@ -136,48 +136,16 @@ Class AzureProvider : TestProvider
...
-						$recheckAgain = $true
-					}
-				}
+		$Timeout = New-Timespan -Minutes ([int]($MaximumCores / 10) + 10)
+		$sw = [diagnostics.stopwatch]::StartNew()
+		foreach ($vmData in $AllVMData) {
+			$vm = Get-AzVM -ResourceGroupName $vmData.ResourceGroupName -Name $vmData.RoleName -Status
+			while (($vm.Statuses[-1].Code -ne "PowerState/running") -and ($sw.elapsed -lt $Timeout)) {
+				Write-LogInfo "VM $($vmData.RoleName) is in $($vm.Statuses[-1].Code) state, still not in running state"
+				$vm = Get-AzVM -ResourceGroupName $vmData.ResourceGroupName -Name $vmData.RoleName -Status
```
Comment (LiliDeng) [inline TestProviders/AzureProvider.psm1:145]: I update RestartAllDeployments, please take a look @harz566 @juhlee-microsoft 
**Total Time (dd:hh:mm) : 0:6:54 => Total Time (dd:hh:mm) : 0:1:42**

Before -
```
[LISAv2 Test Results Summary]
Test Run On           : 04/20/2020 04:11:00
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Initial Kernel Version: 5.0.0-1036-azure
Final Kernel Version  : 5.0.0-1036-azure
Total Test Cases      : 5 (5 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:6:54

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART-MAX-GPU-CUDA                                            PASS                 4.03 
    2 STRESS               STRESSTEST-VERIFY-RESTART                                                         PASS               147.11 
    3 STRESS               STRESSTEST-VERIFY-RESTART-MAX-GPU-GRID                                            PASS                 5.41 
    4 STRESS               STRESSTEST-VERIFY-RESTART-MAX-NVME-DISK                                           PASS               234.05 
    5 STRESS               STRESSTEST-VERIFY-RESTART-MAX-SRIOV-NICS                                          PASS                 8.84 
```
After -
```
[LISAv2 Test Results Summary]
Test Run On           : 04/20/2020 11:12:26
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Initial Kernel Version: 5.0.0-1036-azure
Final Kernel Version  : 5.0.0-1036-azure
Total Test Cases      : 5 (5 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:1:42

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 STRESS               STRESSTEST-VERIFY-RESTART                                                         PASS                29.21 
    2 STRESS               STRESSTEST-VERIFY-RESTART-MAX-GPU-CUDA                                            PASS                 4.44 
    3 STRESS               STRESSTEST-VERIFY-RESTART-MAX-GPU-GRID                                            PASS                 5.56 
    4 STRESS               STRESSTEST-VERIFY-RESTART-MAX-NVME-DISK                                           PASS                43.08 
    5 STRESS               STRESSTEST-VERIFY-RESTART-MAX-SRIOV-NICS                                          PASS                 2.01 
```

Comment (LiliDeng) [reply TestProviders/AzureProvider.psm1:145]:
```
[LISAv2 Test Results Summary]
Test Run On           : 04/21/2020 01:45:53
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Initial Kernel Version: 5.0.0-1036-azure
Final Kernel Version  : 5.0.0-1036-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:27

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 INFINIBAND           INFINIBAND-OPEN-MPI-2VM                                                           PASS                21.18 
	InfiniBand-Verification-1-FirstBoot : eth0 IP : PASS 
	InfiniBand-Verification-1-FirstBoot : IBV_PINGPONG : PASS 
	InfiniBand-Verification-1-FirstBoot : PingPong Intranode : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-MPI1 : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-P2P : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-NBC : PASS 
	InfiniBand-Verification-2-Reboot : eth0 IP : PASS 
	InfiniBand-Verification-2-Reboot : IBV_PINGPONG : PASS 
	InfiniBand-Verification-2-Reboot : PingPong Intranode : PASS 
	InfiniBand-Verification-2-Reboot : IMB-MPI1 : PASS 
	InfiniBand-Verification-2-Reboot : IMB-P2P : PASS 
	InfiniBand-Verification-2-Reboot : IMB-NBC : PASS 
```

File: Libraries/CommonFunctions.psm1 (hunk 9 above)
Comment (LiliDeng) [reply Libraries/CommonFunctions.psm1:2379]: I saw them in debug log.
```
04/20/2020 12:03:55 : [DEBUG] .\Tools\plink.exe -ssh -t -pw ****** -P 22 lisa@52.156.135.161 "echo ****** | sudo -S lspci | grep -i 'Non-Volatile memory controller' | wc -l"
04/20/2020 12:03:57 : [DEBUG] lspci | grep -i 'Non-Volatile memory controller' | wc -l executed successfully in 1.45 seconds.
04/20/2020 12:03:57 : [INFO ] Actual count: 10, expected count: 10
```

**Review by juhlee-microsoft — APPROVED** [review]

**Pull request merged**